clients/spark/core/src/main/scala/io/treeverse/clients/StorageUtils.scala

```diff
@@ -99,66 +100,61 @@ object StorageUtils {
-        logger.info(
-          s"""Bucket "$bucket" is not in region "$region", discovered it in region "$bucketRegion"""
+        initializeS3Client(configuration, credentialsProvider, awsS3ClientBuilder, endpoint)
+      var bucketRegion =
+        try {
+          getAWSS3Region(client, bucket)
+        } catch {
+          case e: Throwable =>
+            logger.info(f"Could not fetch region for bucket ${bucket}", e)
+            ""
+        }
+      if (bucketRegion == "" && region == "") {
+        throw new IllegalArgumentException(
+          s"""Could not fetch region for bucket "$bucket" and no region was provided"""
         )
-        client = initializeS3Client(configuration,
-                                    credentialsProvider,
-                                    AmazonS3ClientBuilder.standard(),
-                                    endpoint,
-                                    bucketRegion
-                                   )
       }
-      client
+      if (bucketRegion == "") {
+        bucketRegion = region
+      }
+      initializeS3Client(configuration,
+                         credentialsProvider,
+                         awsS3ClientBuilder,
+                         endpoint,
+                         bucketRegion
+                        )
     }
 
     private def initializeS3Client(
         configuration: ClientConfiguration,
         credentialsProvider: Option[AWSCredentialsProvider],
         awsS3ClientBuilder: AmazonS3ClientBuilder,
         endpoint: String,
-        region: String
+        region: String = null
     ): AmazonS3 = {
       val builder = awsS3ClientBuilder
         .withClientConfiguration(configuration)
-
       val builderWithEndpoint =
         if (endpoint != null)
           builder.withEndpointConfiguration(
             new AwsClientBuilder.EndpointConfiguration(endpoint, region)
           )
-        else
+        else if (region != null)
           builder.withRegion(region)
-
+        else
+          builder
       val builderWithCredentials = credentialsProvider match {
         case Some(cp) => builderWithEndpoint.withCredentials(cp)
         case None     => builderWithEndpoint
       }
       builderWithCredentials.build
     }
 
-    private def validateClientAndBucketRegionsMatch(client: AmazonS3, bucket: String): Boolean = {
-      try {
-        client.headBucket(new HeadBucketRequest(bucket))
-        true
-      } catch {
-        case e: AmazonServiceException =>
-          logger.info("Bucket \"{}\" isn't reachable with error: {}",
-                      bucket: Any,
-                      e.getMessage: Any
-                     )
-          false
-      }
-    }
-
     private def getAWSS3Region(client: AmazonS3, bucket: String): String = {
-      val bucketRegion = client.getBucketLocation(bucket)
+      var request = new GetBucketLocationRequest(bucket)
+      request = request.withSdkClientExecutionTimeout(TimeUnit.SECONDS.toMillis(15).intValue())
```

Comment:
I think that you can set this value in the `configuration` (`ClientConfiguration`) that is passed to `initializeS3Client`.
It's constructed in the `S3ClientBuilder` files.

Reply:
Since this timeout applies to the entirety of an SDK operation (and not specific API requests), I don't want to take the risk of applying it globally.